src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs

```diff
@@ -0,0 +1,337 @@
+        /// </summary>
+        /// <param name="buildId">Maestro build id.</param>
+        /// <param name="channelId">Maestro channel id.</param>
+        /// <returns></returns>
+        public async Task RunAssociatedReleasePipelinesAsync(int buildId, int channelId, CancellationToken cancellationToken)
+        {
+            Logger.LogInformation($"Starting release pipeline for {buildId} in {channelId}");
+            Build build = await Context.Builds
+                .Where(b => b.Id == buildId).FirstAsync();
+
+            Channel channel = await Context.Channels
+                .Where(ch => ch.Id == channelId)
+                .Include(ch => ch.ChannelReleasePipelines)
+                .ThenInclude(crp => crp.ReleasePipeline)
+                .FirstAsync();
+
+            // If something use the old API version we won't have this information available.
+            // This will also be the case if something adds an existing build (created using
+            // the old API version) to a channel
+            if (build.AzureDevOpsBuildId == null)
+            {
+                Logger.LogInformation($"barBuildInfo.AzureDevOpsBuildId is null for BAR Build.Id {build.Id}.");
+                return;
+            }
+
+            if (channel.ChannelReleasePipelines?.Any() != true)
+            {
+                Logger.LogInformation($"Channel {channel.Id}, which build with BAR ID {build.Id} is attached to, doesn't have an associated publishing pipeline.");
+                return;
+            }
+
+            AzureDevOpsClient azdoClient = await GetAzureDevOpsClientForAccount(build.AzureDevOpsAccount);
+
+            var azdoBuild = await azdoClient.GetBuildAsync(
+                build.AzureDevOpsAccount,
+                build.AzureDevOpsProject,
+                build.AzureDevOpsBuildId.Value);
+
+            var runningPipelines =
+                await StateManager.GetOrAddAsync<IReliableDictionary<int, IList<ReleasePipelineStatusItem>>>(RunningPipelineDictionaryName);
+            List<ReleasePipelineStatusItem> releaseList = new List<ReleasePipelineStatusItem>();
+
+            foreach (ChannelReleasePipeline pipeline in channel.ChannelReleasePipelines)
+            {
+                try
+                {
+                    string organization = pipeline.ReleasePipeline.Organization;
+                    string project = pipeline.ReleasePipeline.Project;
+                    int pipelineId = pipeline.ReleasePipeline.PipelineIdentifier;
+
+                    AzureDevOpsReleaseDefinition pipeDef = await azdoClient.GetReleaseDefinitionAsync(organization, project, pipelineId);
+                    pipeDef = await azdoClient.RemoveAllArtifactSourcesAsync(organization, project, pipeDef);
+
+                    pipeDef = await azdoClient.AddArtifactSourceAsync(organization, project, pipeDef, azdoBuild);
+
+                    int releaseId = await azdoClient.StartNewReleaseAsync(organization, project, pipeDef, build.Id);
```

Comment:
Are these command safe to be run in paralle? I can't really tell what they are doing, but I think they are messing around with the artifact sources of a pipeline definiton, which might mean if 2 things are looking at the same definition, they'll be adding and removing stuff, causing weird conflicts.

Do I misunderstand this? The ReleasePipelineRunner is set to 3 instances, so it might be fighting with itself here. We will probably need some sort of "I'm locking this pipeline definition, no one is allowed to touch it" type thing.

Reply:
Discussed offline. Right now they won't run in parallel. Once this goes in I'll create a follow up issue to consider improving this part.

Reply:
RunAsync only runs in the primary instance. This is a valid concern if we ever want to do this in parallel. Once this goes in I'll file an issue so we keep this limitation in mind and find alternatives.